staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go

```diff
@@ -82,29 +74,92 @@ func init() {
 	// we need to explicitly register these metrics to our global registry here.
 	// For reference: https://github.com/kubernetes/kubernetes/pull/81387
 	legacyregistry.RawMustRegister(grpcprom.DefaultClientMetrics)
-	dbMetricsMonitors = make(map[string]struct{})
+}
+
+type activeETCD3Client struct {
+	client *clientv3.Client
+	cancel func() // Cancel background goroutines (compaction etc).
+	refs   int
+
+	compactionInterval time.Duration
+	metricPollInterval time.Duration
+}
+
+type etcd3ClientCache struct {
+	mx     sync.Mutex
+	active map[string]*activeETCD3Client
+}
+
+func newETCD3ClientCache() *etcd3ClientCache {
+	return &etcd3ClientCache{active: make(map[string]*activeETCD3Client)}
+}
+
+func etcd3ClientCacheKey(tc storagebackend.TransportConfig) string {
+	return fmt.Sprintf("%v", tc) // gives: {[server1 server2] keyFile certFile caFile}
+}
+
+// For returns a compacted etcd v3 client for the supplied transport config. One
+// client and compactor are started for each transport. Callers must call the
+// returned destroy function once they are done with the client.
+func (e *etcd3ClientCache) For(tc storagebackend.TransportConfig, compactionInterval, metricPollInterval time.Duration) (*clientv3.Client, error) {
+	e.mx.Lock()
+	defer e.mx.Unlock()
+
+	k := etcd3ClientCacheKey(tc)
+	if a, ok := e.active[k]; ok {
+		// We want to compact and poll at the shortest interval specified by any caller.
+		if a.compactionInterval > compactionInterval || a.metricPollInterval > metricPollInterval {
```

Comment:
what if one is lower and other greater, you'll update both

Reply:
True - that's a bit of a shame I would have liked to avoid the complexity of handling them both separately.

Reply:
Just pushed a commit to address this.